### PR TITLE
Reduce sunpy.map import time

### PR DESCRIPTION
**Pull request by dstansby**
xref #3448. This reduces import time from 3.8s to 3.0s for me locally.

### CODE REVIEW

**Comment by nabobalis** [comment]
Any more changes you want to make?

**Comment by dstansby** [comment]
These are the only simple ones I can see. Will add a changelog in a bit.

**Comment by nabobalis** [comment]
> These are the only simple ones I can see. Will add a changelog in a bit.

How much effort is to to remake the import time figure?

**Comment by Cadair** [comment]
I wouldn't backport this personally, mainly as we are so close to 1.1. We can have "Import Performance Improvements" in the 1.1 changelog :grin: 

I am still unsure if I like imports in functions, but it's worth it for a 20% import time improvement.

File: sunpy/map/mapbase.py [hunk 1]
```diff
@@ -25,7 +25,6 @@
 from sunpy.sun import constants
 from sunpy.coordinates import sun
 from sunpy.time import parse_time, is_time
-from sunpy.image.transform import affine_transform
 from sunpy.image.resample import reshape_image_to_4d_superpixel
```
Comment (Cadair) [inline sunpy/map/mapbase.py]: Do these imports not pull in `affine_transform` anyway?

Comment (dstansby) [reply sunpy/map/mapbase.py]: Nope; what moving this import really does is avoid having to import skimage.

**Comment by nabobalis** [comment]
> I wouldn't backport this personally, mainly as we are so close to 1.1. We can have "Import Performance Improvements" in the 1.1 changelog 
> 
> I am still unsure if I like imports in functions, but it's worth it for a 20% import time improvement.

Fine. I removed the backport.

I also can't say I am a fan of imports in functions but since the import time is slow, it's a sacrifice worth making.

File: sunpy/map/mapbase.py (hunk 1 above)
Comment (Cadair) [reply sunpy/map/mapbase.py]: I figured, I was just surprised.

**Comment by dstansby** [comment]
> > These are the only simple ones I can see. Will add a changelog in a bit.
> 
> How much effort is to to remake the import time figure?

Very little, here's the new one:

Screenshot 2019-11-12 at 16 12 46

**Comment by nabobalis** [comment]
Hmm. Could be other places to hit but nothing simple.

**Comment by dstansby** [comment]
Yes, quite a lot of the potential places for improvements (e.g. pandas, matplotlib) need to be imported to register various external classes with an internal SunPy registry (e.g. for time conversion), and I'm not sure if there's an way to get around that.

**Comment by Cadair** [comment]
I got this down to sub 1s on Python 3.8

Screen Shot 2019-11-12 at 16 33 16

It does have one proper weird work around though, which I will open as a PR to this branch, because it's weird.

**Comment by nabobalis** [comment]
What on earth did you do?

**Comment by Cadair** [comment]
https://github.com/dstansby/sunpy/pull/2

I don't think the times are comparable, I think my PC might just be a bit faster than Davids. Also it's a very clean Python 3.8 env. I only ran (`pip install -e .[all]`).

**Comment by Cadair** [comment]
For comparison without the patch I get this:
Screen Shot 2019-11-12 at 16 40 34

**Comment by nabobalis** [comment]
> [dstansby#2](https://github.com/dstansby/sunpy/pull/2)
> 
> I don't think the times are comparable, I think my PC might just be a bit faster than Davids. Also it's a very clean Python 3.8 env. I only ran (`pip install -e .[all]`).

Isn't all everything but testing and docs deps? I would have thought it would import all of the optional deps?

**Comment by Cadair** [comment]
The only thing is `dask` which we import if it's installed. Otherwise I think it will be ok.

**Comment by nabobalis** [comment]
Uh oh.

File: sunpy/util/types.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+"""
+A collection of ABCs to use to shortcut having to import things.
+"""
+import abc
+
```
Comment (Cadair) [inline sunpy/util/types.py]:
```suggestion


```

File: sunpy/util/types.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+class DatabaseEntryType(metaclass=abc.ABCMeta):
+    """
+    A Base class for sunpy.database.DatabaseEntry.
```
Comment (Cadair) [inline sunpy/util/types.py]:
```suggestion
    A base class for `sunpy.database.tables.DatabaseEntry`.
```

File: sunpy/util/types.py (hunk 2 above)
Comment (nabobalis) [reply sunpy/util/types.py]: You don't need to import abc for metaclasses to work?

Comment (Cadair) [reply sunpy/util/types.py]: The suggestion is adding an extra line for the pep8 checker.

**Comment by nabobalis** [comment]
The ABC patch broke the entire suite. 

```
class DatabaseEntry(DatabaseEntryType, Base):

TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

Not a clue how to fix it.

File: changelog/3495.feature.rst [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Made a couple of module imports lazy to reduce the import time of sunpy.map by
+~20%.
```
Comment (ayshih) [inline changelog/3495.feature.rst]: On my machine, the total time improvement is now ~40% (from ~6 seconds to ~3.5 seconds).

File: sunpy/util/types.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+"""
+A collection of ABCs to use to shortcut having to import things.
```
Comment (ayshih) [inline sunpy/util/types.py]: I think this docstring should be a little more descriptive, including elaborating on "things".

File: changelog/3495.feature.rst (hunk 4 above)
Comment (nabobalis) [reply changelog/3495.feature.rst]: Go with the worst case? Or we just remove the % and say its just faster.

Comment (ayshih) [reply changelog/3495.feature.rst]: Oh, I believe the "~20%" was estimated prior to the `database` improvement, so I suspect that ~40% is more accurate for everyone.

I like having a value in the changelog entry.

**Pull request merged**